### ISSUE TITLE
Moving a file out of a share corrupts its versions

### ISSUE THREAD

**Issue by phil-davis** [owncloud/encryption#321]
With Master Key encryption
Against core master:
https://drone.owncloud.com/owncloud/encryption/2260/51/18

Against core 10.9.1RC1
https://drone.owncloud.com/owncloud/encryption/2260/91/18

```
Scenario: enable file versioning and check the history of changes from multiple users while moving file in/out of a subfolder                      # /var/www/owncloud/testrunner/tests/acceptance/features/apiVersions/fileVersionAuthor.feature:75
    Given user "Alice" has created folder "/test"                                                                                                    # FeatureContext::userHasCreatedFolder()
    And group "grp1" has been created                                                                                                                # FeatureContext::groupHasBeenCreated()
    And user "Alice" has been added to group "grp1"                                                                                                  # FeatureContext::userHasBeenAddedToGroup()
    And user "Brian" has been added to group "grp1"                                                                                                  # FeatureContext::userHasBeenAddedToGroup()
    And user "Carol" has been added to group "grp1"                                                                                                  # FeatureContext::userHasBeenAddedToGroup()
    And user "Alice" has shared folder "/test" with group "grp1"                                                                                     # FeatureContext::userHasSharedFileWithGroupUsingTheSharingApi()
    And user "Alice" has uploaded file with content "uploaded content alice" to "/test/textfile0.txt"                                                # FeatureContext::userHasUploadedAFileWithContentTo()
    And user "Brian" has uploaded file with content "uploaded content brian" to "/test/textfile0.txt"                                                # FeatureContext::userHasUploadedAFileWithContentTo()
    And user "Carol" has uploaded file with content "uploaded content carol" to "/test/textfile0.txt"                                                # FeatureContext::userHasUploadedAFileWithContentTo()
    When user "Brian" moves file "/test/textfile0.txt" to "/textfile0.txt" using the WebDAV API                                                      # FeatureContext::userMovesFileUsingTheAPI()
    And user "Brian" uploads file with content "uploaded content brian after moving file outside subfolder" to "/textfile0.txt" using the WebDAV API # FeatureContext::userUploadsAFileWithContentTo()
    And user "Brian" moves file "/textfile0.txt" to "/test/textfile0.txt" using the WebDAV API                                                       # FeatureContext::userMovesFileUsingTheAPI()
    Then as "Alice" file "/test/textfile0.txt" should exist                                                                                          # FeatureContext::asFileOrFolderShouldExist()
    When user "Alice" gets the number of versions of file "/test/textfile0.txt"                                                                      # FilesVersionsContext::userGetsFileVersions()
    Then the number of versions should be "3"                                                                                                        # FeatureContext::theNumberOfVersionsShouldBe()
    And the content of version index "1" of file "/test/textfile0.txt" for user "Alice" should be "uploaded content carol"                           # FilesVersionsContext::theContentOfVersionIndexOfFileForUserShouldBe()
      HTTP status code 500 is not the expected value 200
      Failed asserting that 500 matches expected '200'.
```

The fail happens at the same step in both test pipelines.

**Comment by phil-davis** [comment]
server log: https://drone.owncloud.com/owncloud/encryption/2260/91/13
```
{"reqId":"02gdqJvtdfKeTLqYmsIP","level":4,"time":"2022-01-07T01:12:36+00:00","remoteAddr":"192.168.3.7","user":"Alice","app":"webdav",
"method":"GET","url":"\/remote.php\/dav\/meta\/2147483744\/v\/1641517952",
"message":"Exception: Bad Signature: {\"Exception\":\"OC\\\\HintException\",\"Message\":\"Bad Signature\",\"Code\":0,\"Trace\":\"
#0 \\\/var\\\/www\\\/owncloud\\\/server\\\/apps\\\/encryption\\\/lib\\\/Crypto\\\/Crypt.php(504): OCA\\\\Encryption\\\\Crypto\\\\Crypt->checkSignature()\\n
#1 \\\/var\\\/www\\\/owncloud\\\/server\\\/apps\\\/encryption\\\/lib\\\/Crypto\\\/Encryption.php(377): OCA\\\\Encryption\\\\Crypto\\\\Crypt->symmetricDecryptFileContent()\\n
#2 \\\/var\\\/www\\\/owncloud\\\/server\\\/lib\\\/private\\\/Files\\\/Stream\\\/Encryption.php(478): OCA\\\\Encryption\\\\Crypto\\\\Encryption->decrypt(*** sensitive parameters replaced ***)\\n
#3 \\\/var\\\/www\\\/owncloud\\\/server\\\/lib\\\/private\\\/Files\\\/Stream\\\/Encryption.php(294): OC\\\\Files\\\\Stream\\\\Encryption->readCache()\\n
#4 [internal function]: OC\\\\Files\\\\Stream\\\\Encryption->stream_read()\\n
#5 \\\/var\\\/www\\\/owncloud\\\/server\\\/apps\\\/files_external\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/Wrapper.php(55): fread()\\n
#6 \\\/var\\\/www\\\/owncloud\\\/server\\\/apps\\\/files_external\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/CallbackWrapper.php(96): Icewind\\\\Streams\\\\Wrapper->stream_read()\\n
#7 [internal function]: Icewind\\\\Streams\\\\CallbackWrapper->stream_read()\\n
#8 \\\/var\\\/www\\\/owncloud\\\/server\\\/lib\\\/composer\\\/sabre\\\/http\\\/lib\\\/Sapi.php(112): stream_copy_to_stream()\\n
#9 \\\/var\\\/www\\\/owncloud\\\/server\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(490): Sabre\\\\HTTP\\\\Sapi::sendResponse()\\n
#10 \\\/var\\\/www\\\/owncloud\\\/server\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n
#11 \\\/var\\\/www\\\/owncloud\\\/server\\\/apps\\\/dav\\\/lib\\\/Server.php(345): Sabre\\\\DAV\\\\Server->start()\\n
#12 \\\/var\\\/www\\\/owncloud\\\/server\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n
#13 \\\/var\\\/www\\\/owncloud\\\/server\\\/remote.php(165): require_once('\\\/var\\\/www\\\/ownclo...')\\n
#14 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/server\\\/apps\\\/encryption\\\/lib\\\/Crypto\\\/Crypt.php\",\"Line\":528}"}
```

**Comment by phil-davis** [comment]
It fails at a point that is not directly related to the version author, but is a problem with versions.

- Alice shares a folder to Brian (and others...)
- The folder contains a file
- Brian moves the file out of the share to his own private root folder
- Brian changes the content
- Brian moves the file back into the share

At this point the test starts doing various checks of the outcome. The first thing it does is download a version of the file, and check the content. That gets a 500 status, and the server log the Bad Signature exception.

**Comment by jnweiger** [comment]
Repoduced the same a bt simpler manually,:

* user01 shares a folder to user02
* user02 edits the file in the folder with text_editor.
* user02 moves the file out of the share to his own private root folder
* user02 tries to download a previous version. nothing happens on the UI. A "Bad Signature" error is logged:

```
{"reqId":"Ydg9JF9oRXOuw4621P8haAAAAAc","level":4,"time":"2022-01-07T13:16:20+00:00","remoteAddr":"2.247.253.63","user":"user02","app":"webdav","method":"GET
","url":"\/remote.php\/dav\/meta\/542\/v\/1641561061","message":"Exception: Bad Signature: {\"Exception\":\"OC\\\\HintException\",\"Message\":\"Bad Signatur
e\",\"Code\":0,\"Trace\":\"#0 \\\/var\\\/www\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/Crypto\\\/Crypt.php(481): OCA\\\\Encryption\\\\Crypto\\\\Crypt->che
ckSignature()\\n#1 \\\/var\\\/www\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/Crypto\\\/Encryption.php(373): OCA\\\\Encryption\\\\Crypto\\\\Crypt->symmetric
DecryptFileContent()\\n#2 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Stream\\\/Encryption.php(478): OCA\\\\Encryption\\\\Crypto\\\\Encryption->decrypt(*** sensitive parameters replaced ***)\\n#3 \\\/var\\\/www\\\/owncloud\\\/lib\\\/private\\\/Files\\\/Stream\\\/Encryption.php(294): OC\\\\Files\\\\Stream\\\\Encryption->readCache()\\n#4 [internal function]: OC\\\\Files\\\\Stream\\\\Encryption->stream_read()\\n#5 \\\/var\\\/www\\\/owncloud\\\/apps\\\/files_external\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/Wrapper.php(55): fread()\\n#6 \\\/var\\\/www\\\/owncloud\\\/apps\\\/files_external\\\/3rdparty\\\/icewind\\\/streams\\\/src\\\/CallbackWrapper.php(96): Icewind\\\\Streams\\\\Wrapper->stream_read()\\n#7 [internal function]: Icewind\\\\Streams\\\\CallbackWrapper->stream_read()\\n#8 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/http\\\/lib\\\/Sapi.php(112): stream_copy_to_stream()\\n#9 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(490): Sabre\\\\HTTP\\\\Sapi::sendResponse()\\n#10 \\\/var\\\/www\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(253): Sabre\\\\DAV\\\\Server->invokeMethod()\\n#11 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(345): Sabre\\\\DAV\\\\Server->start()\\n#12 \\\/var\\\/www\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#13 \\\/var\\\/www\\\/owncloud\\\/remote.php(165): require_once('\\\/var\\\/www\\\/ownclo...')\\n#14 {main}\",\"File\":\"\\\/var\\\/www\\\/owncloud\\\/apps\\\/encryption\\\/lib\\\/Crypto\\\/Crypt.php\",\"Line\":504}"}

```

* user02 restores a version and tries to edit again
![image](https://user-images.githubusercontent.com/1108546/148548535-15b5f419-9bf1-4473-b706-1e2db5b1d8bc.png)

If user02 does not restore (an already broken) version, he can edit the file in the new location.
The version lists then has new versions that can be downloaded, and older versions that fail with Bad Signature.

**Comment by JammingBen** [comment]
Reproduced, this is not a regression due to 10.9. The error exists with 10.8 as well. The test itself is new, that's why it pops up now.

Needs fixing of course, but probably not in 10.9.1.

**Comment by jnweiger** [comment]
Reproduced with 10.9.1RC2 exactly as described in https://github.com/owncloud/encryption/issues/321#issuecomment-1007402941

**Comment by phil-davis** [comment]
Note: the test scenario has been skipped when running with encryption, in order to let the rest of encryption CI run and pass.

https://github.com/owncloud/core/pull/39681

When fixing this issue, it should be easy to put a similar scenario in the `apiEncryption` test suite and have it fail, then fix the code and have it pass. When the problem is fixed, make sure to remove the `@skipOnEncryption` tag from the scenario in core.

**Comment by mrow4a** [comment]
same problem as in https://github.com/owncloud/core/issues/39664#issuecomment-1024160655, if we fix that one we will fix this one

**Comment by jvillafanez** [comment]
The problem seems to be really bad if my suspicions are right.
First, there are some assumptions:
* The fileKey for the file is shared with all the versions
* Versions are copied without any change of the content of the file

Most of the problems so far came from the second point: we were increasing the encrypted value in the DB for the versions, causing the signature check to fail. Adjusting the encrypted value to the right one or bypassing the signature check can fix this issue because the password file is still unchanged, so both the file and the versions could be decrypted.

The problem here is that, when the file is moved out of the share, the password file changes (and the password itself likely changes too). What happens, I guess, is:
1. File is moved out of the share
2. Password file is regenerated for the target file
3. Versions are moved without any change

Since the password has been changed, the versions that have been moved along with the file are broken and cannot be decrypted. Note that this is due to the wrong password, not a mismatched encrypted version.

**Comment by jvillafanez** [comment]
The problem seems to be around https://github.com/owncloud/core/blob/master/lib/private/Files/Storage/Wrapper/Encryption.php#L779-L787
Although there is only one mount point in this particular case, the problem is that the keys are moved from "/user2/files/test//test1.txt" (there is a double "/" somehow, but it shouldn't matter) to "/user2/files/test1.txt", however, there are no keys in the source because the file is shared in the keys belong to the owner ("admin" in my case, not "user2"). This means that the keys aren't copied and new ones are generated, which is what breaks the versions.

**Comment by jvillafanez** [comment]
https://github.com/owncloud/core/pull/39829 should fix this issue.

**Comment by jvillafanez** [comment]
https://github.com/owncloud/core/pull/39829 merged. Closing.

**Comment by jnweiger** [comment]
Confirmed fixed in 10.10.0 RC2
